File: item-list/src/test/java/com/gurunars/item_list/DifferTest.java
package com.gurunars.item_list;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.junit.Assert.assertEquals;


public class DifferTest {

    private Differ<AnimalItem> differ = new Differ<>();

    @Test
    public void itemCreate_shouldProduceSingleItemListWithCreation() throws Exception {
        assertEquals(
                Collections.singletonList(
                        new ChangeCreate<>(new ItemHolder<>(new AnimalItem(7, 1)), 6, 6)),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1),
                                new AnimalItem(7, 1)
                        )))
        );

    }

    @Test
    public void itemDelete_shouldProduceSingleItemListWithRemoval() throws Exception {
        assertEquals(
                Collections.singletonList(
                        new ChangeDelete<>(new ItemHolder<>(new AnimalItem(3, 1)), 2, 2)),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )),
                                ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )))
        );

    }

    @Test
    public void moveUp_shouldProduceSingleItemList() throws Exception {
        assertEquals(
                Collections.singletonList(
                        new ChangeMove<>(new ItemHolder<>(new AnimalItem(5, 1)), 4, 1)),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(5, 1),  // moved up
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(6, 1)
                        )))
        );

    }

    @Test
    public void moveDown_shouldProduceSingleItemList() throws Exception {
        assertEquals(
                Collections.singletonList(
                        new ChangeMove<>(new ItemHolder<>(new AnimalItem(2, 1)), 1, 4)),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(2, 1), // moved down
                                new AnimalItem(6, 1)
                        )))
        );
    }

    @Test
    public void itemEdit_shouldProduceSingleItemListWithUpdate() throws Exception {
        assertEquals(
                Collections.singletonList(
                        new ChangeUpdate<>(new ItemHolder<>(new AnimalItem(3, 2)), 2, 2)),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 2), // updated
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1)
                        )))
        );

    }

    @Test
    public void simpleDiff_isCorrect() throws Exception {

        assertEquals(
                Arrays.asList(
                        new ChangeDelete<>(new ItemHolder<>(new AnimalItem(5, 1)), 4, 4),
                        new ChangeDelete<>(new ItemHolder<>(new AnimalItem(2, 1)), 1, 1),
                        new ChangeCreate<>(new ItemHolder<>(new AnimalItem(7, 2)), 0, 0),
                        new ChangeCreate<>(new ItemHolder<>(new AnimalItem(9, 1)), 4, 4),
                        new ChangeMove<>(new ItemHolder<>(new AnimalItem(1, 1)), 1, 3),
                        new ChangeMove<>(new ItemHolder<>(new AnimalItem(3, 1)), 1, 2),
                        new ChangeUpdate<>(new ItemHolder<>(new AnimalItem(1, 2)), 3, 3),
                        new ChangeUpdate<>(new ItemHolder<>(new AnimalItem(3, 2)), 2, 2)
                ),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(7, 2),
                                new AnimalItem(4, 1),
                                new AnimalItem(3, 2),
                                new AnimalItem(1, 2),
                                new AnimalItem(9, 1)
                        )))
        );

    }

    @Test
    public void tooManyMoves_resultInComplexPermutation() {

        assertEquals(
                Collections.singletonList(new ChangeComplexPermutation<>(1, ItemHolder.wrap(Arrays.asList(
                        new AnimalItem(7, 1),
                        new AnimalItem(5, 1),
                        new AnimalItem(8, 1),
                        new AnimalItem(6, 1),
                        new AnimalItem(4, 1),
                        new AnimalItem(2, 1),
                        new AnimalItem(3, 1)
                )))),
                differ.apply(
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(6, 1),
                                new AnimalItem(7, 1),
                                new AnimalItem(8, 1),
                                new AnimalItem(9, 1)
                        )),
                        ItemHolder.wrap(Arrays.asList(
                                new AnimalItem(1, 1),
                                new AnimalItem(7, 1),
                                new AnimalItem(5, 1),
                                new AnimalItem(8, 1),
                                new AnimalItem(6, 1),
                                new AnimalItem(4, 1),
                                new AnimalItem(2, 1),
                                new AnimalItem(3, 1),
                                new AnimalItem(9, 1)
                        )))
        );

    }

}

File: item-list/src/test/java/com/gurunars/item_list/DifferTest.kt
package com.gurunars.item_list

import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*


class DifferTest {

    private val differ = Differ<AnimalItem>()

    @Test
    @Throws(Exception::class)
    fun itemCreate_shouldProduceSingleItemListWithCreation() {
        assertEquals(
                listOf(ChangeCreate(AnimalItem(7, 1), 6, 6)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1),
                                AnimalItem(7, 1)
                        ))
        )

    }

    @Test
    @Throws(Exception::class)
    fun itemDelete_shouldProduceSingleItemListWithRemoval() {
        assertEquals(
                listOf(ChangeDelete(AnimalItem(3, 1), 2, 2)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ))
        )

    }

    @Test
    @Throws(Exception::class)
    fun moveUp_shouldProduceSingleItemList() {
        assertEquals(
                listOf(ChangeMove(AnimalItem(5, 1), 4, 1)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(5, 1), // moved up
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(6, 1)
                        ))
        )

    }

    @Test
    @Throws(Exception::class)
    fun moveDown_shouldProduceSingleItemList() {
        assertEquals(
                listOf(ChangeMove(AnimalItem(2, 1), 1, 4)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(2, 1), // moved down
                                AnimalItem(6, 1)
                        ))
        )
    }

    @Test
    @Throws(Exception::class)
    fun itemEdit_shouldProduceSingleItemListWithUpdate() {
        assertEquals(
                listOf(ChangeUpdate(AnimalItem(3, 2), 2, 2)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 2), // updated
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1)
                        ))
        )

    }

    @Test
    @Throws(Exception::class)
    fun simpleDiff_isCorrect() {

        assertEquals(
                Arrays.asList(
                        ChangeDelete(AnimalItem(5, 1), 4, 4),
                        ChangeDelete(AnimalItem(2, 1), 1, 1),
                        ChangeCreate(AnimalItem(7, 2), 0, 0),
                        ChangeCreate(AnimalItem(9, 1), 4, 4),
                        ChangeMove(AnimalItem(1, 1), 1, 3),
                        ChangeMove(AnimalItem(3, 1), 1, 2),
                        ChangeUpdate(AnimalItem(1, 2), 3, 3),
                        ChangeUpdate(AnimalItem(3, 2), 2, 2)
                ),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(7, 2),
                                AnimalItem(4, 1),
                                AnimalItem(3, 2),
                                AnimalItem(1, 2),
                                AnimalItem(9, 1)
                        ))
        )

    }

    @Test
    fun tooManyMoves_resultInComplexPermutation() {

        assertEquals(
                listOf(ChangeComplexPermutation(1, Arrays.asList(
                        AnimalItem(7, 1),
                        AnimalItem(5, 1),
                        AnimalItem(8, 1),
                        AnimalItem(6, 1),
                        AnimalItem(4, 1),
                        AnimalItem(2, 1),
                        AnimalItem(3, 1)
                ))),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 1),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1),
                                AnimalItem(7, 1),
                                AnimalItem(8, 1),
                                AnimalItem(9, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(7, 1),
                                AnimalItem(5, 1),
                                AnimalItem(8, 1),
                                AnimalItem(6, 1),
                                AnimalItem(4, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(9, 1)
                        ))
        )

    }

    @Test
    fun movesDeltaOne() {
        assertEquals(
                listOf(ChangeMove(AnimalItem(4, 2), 3, 4)),
                differ.apply(
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(4, 2),
                                AnimalItem(5, 1),
                                AnimalItem(6, 1),
                                AnimalItem(7, 1),
                                AnimalItem(8, 1),
                                AnimalItem(9, 1)
                        ),
                        Arrays.asList(
                                AnimalItem(1, 1),
                                AnimalItem(2, 1),
                                AnimalItem(3, 1),
                                AnimalItem(5, 1),
                                AnimalItem(4, 2),
                                AnimalItem(6, 1),
                                AnimalItem(7, 1),
                                AnimalItem(8, 1),
                                AnimalItem(9, 1)
                        ))
        )
    }

}